agent/tests/test_user_instrument_hashmap.c

```diff
@@ -272,81 +362,229 @@ static void test_zend_string_hash_after_set_before_get() {
+  rc = nr_php_wraprec_hashmap_get_into(h, &user_function_with_scope,
+                                       &wraprec_found);
+  tlib_pass_if_int_equal("can find named function w/scope after hash reset", 1,
+                         rc);
+  tlib_pass_if_ptr_equal("can find named function w/scope after hash reset",
+                         &wr3, wraprec_found);
 
   zend_function_copy = user_function_with_scope;
   wraprec_found = NULL;
   rc = nr_php_wraprec_hashmap_get_into(h, &zend_function_copy, &wraprec_found);
-  tlib_pass_if_int_equal("can find named function w/scope by zend_function copy", 1, rc);
-  tlib_pass_if_ptr_equal("can find named function w/scope by zend_function copy", &wr3, wraprec_found);
+  tlib_pass_if_int_equal(
+      "can find named function w/scope by zend_function copy", 1, rc);
+  tlib_pass_if_ptr_equal(
+      "can find named function w/scope by zend_function copy", &wr3,
+      wraprec_found);
 
   nr_php_wraprec_hashmap_destroy(&h);
 
   mock_zend_function_destroy(&user_closure);
   mock_zend_function_destroy(&user_function);
   mock_zend_function_destroy(&user_function_with_scope);
 }
+
+static void test_wraprec_hashmap_two_functions() {
+#define FILE_1_NAME "/some/random/path/to/a_file.php"
+#define FILE_2_NAME "/some/random/path/to/b_file.php"
+#define LINENO_BASE 10
+#define SCOPE_1_NAME "a_scope"
+#define FUNC_1_NAME "a_function"
+#define SCOPE_2_NAME "b_scope"
+#define FUNC_2_NAME "b_function"
+
+  zend_function zf1 = {0};
+  zend_function zf2 = {0};
+  nruserfn_t wr1 = {0}, wr2 = {0};
+  int rc = 0;
+  nruserfn_t* wraprec_found = NULL;
+  nr_php_wraprec_hashmap_t* h = NULL;
+  nr_php_wraprec_hashmap_stats_t s = {};
+
+  h = nr_php_wraprec_hashmap_create_buckets(16, reset_wraprec);
+  tlib_fail_if_null("hashmap created", h);
+
+  /* A function */
+  mock_user_function_with_scope(&zf1, FILE_1_NAME, LINENO_BASE, SCOPE_1_NAME,
+                                FUNC_1_NAME);
+  nr_php_wraprec_hashmap_update(h, &zf1, &wr1);
+
+  /* A function with the same in the same file but different scope */
+  mock_user_function_with_scope(&zf2, FILE_1_NAME, LINENO_BASE, SCOPE_2_NAME,
+                                FUNC_1_NAME);
+  nr_php_wraprec_hashmap_update(h, &zf2, &wr2);
+
+  wraprec_found = NULL;
+  rc = nr_php_wraprec_hashmap_get_into(h, &zf1, &wraprec_found);
+  tlib_pass_if_int_equal(
+      "Two functions with the same in the same file but different scope are "
+      "stored separetely",
+      1, rc);
+  tlib_pass_if_ptr_equal(
+      "Two functions with the same in the same file but different scope are "
+      "stored separetely",
+      &wr1, wraprec_found);
+
+  wraprec_found = NULL;
+  rc = nr_php_wraprec_hashmap_get_into(h, &zf2, &wraprec_found);
+  tlib_pass_if_int_equal(
+      "Two functions with the same in the same file but different scope are "
+      "stored separetely",
+      1, rc);
+  tlib_pass_if_ptr_equal(
+      "Two functions with the same in the same file but different scope are "
```

Comment:
```suggestion
      "Two functions with the same name in the same file but different scope are "
```

Reply:
More like this below I didn't mark them all...

Reply:
@mfulb Please see 050d68d9531f03d123725e6e35dbd73e97c3c012. Is this approach acceptable?